engine/src/xetex_xetex0.rs

```diff
@@ -10026,21 +10031,19 @@ pub(crate) unsafe fn hpack(mut popt: Option<usize>, mut w: scaled_t, m: PackMode
                 TxtNode::WhatsIt(w) => match w {
                     WhatsIt::NativeWord(mut p_nw) => {
                         let mut k = if q != r.ptr() + 5 {
-                            if NODE_type(q) == TextNode::Disc.into() {
-                                let q = Discretionary(q);
-                                q.replace_count() as i32
-                            } else {
-                                0
+                            match Node::from(q) {
+                                Node::Text(TxtNode::Disc(d)) => d.replace_count() as i32,
+                                _ => 0,
                             }
                         } else {
                             0
                         };
                         while llist_link(q) != Some(p) {
                             k -= 1;
                             q = *LLIST_link(q) as usize;
-                            if NODE_type(q) == TextNode::Disc.into() {
-                                let q = Discretionary(q);
-                                k = q.replace_count() as i32
+                            match Node::from(q) {
+                                Node::Text(TxtNode::Disc(d)) => k = d.replace_count() as i32,
+                                _ => {}
                             }
```

Comment:
I think there is bug in original code.
Sometimes `q` can be `Char` with code `7` that matches the `TextNode::Disc`. So you need first check it with `is_char_node`.
See `1702.02157`.
cc @pkgw

Reply:
Thanks for flagging this! Did you ever boil it down to a small reproduction test case? I see that 1702.02157 triggers this codepath but it's not obvious to me how the issue manifests in the final output.